src/main/scala/org/constellation/gossip/sampling/PartitionerPeerSampling.scala

```diff
@@ -0,0 +1,70 @@
+package org.constellation.gossip.sampling
+
+import cats.effect.Concurrent
+import cats.effect.concurrent.Ref
+import cats.implicits._
+import io.chrisdavenport.fuuid.FUUID
+import io.chrisdavenport.log4cats.SelfAwareStructuredLogger
+import io.chrisdavenport.log4cats.slf4j.Slf4jLogger
+import org.constellation.domain.trust.TrustDataInternal
+import org.constellation.p2p.Cluster
+import org.constellation.schema.Id
+import org.constellation.trust.TrustManager
+import org.constellation.util.Partitioner._
+
+class PartitionerPeerSampling[F[_]: Concurrent](selfId: Id, cluster: Cluster[F], trustManager: TrustManager[F])
+    extends PeerSampling[F] {
+  private val partitionCache: Ref[F, List[GossipPath]] = Ref.unsafe(List.empty[GossipPath])
+  private val trustDataInternalCache: Ref[F, List[TrustDataInternal]] = Ref.unsafe(List.empty[TrustDataInternal])
+
+  private val logger: SelfAwareStructuredLogger[F] = Slf4jLogger.getLogger[F]
+
+  override def selectPaths: F[List[GossipPath]] =
+    cachedPartitionsCoverAllPeers.ifM(
+      logger.debug("Cached partitions cover all the peers. Returning.") >> partitionCache.get,
+      logger
+        .debug("Some nodes are missing in cached partitions. Repartitioning.") >> repartitionWithDefaults() >> partitionCache.get
+    )
+
+  def repartition(selfTdi: TrustDataInternal, tdi: List[TrustDataInternal]): F[Unit] =
+    for {
+      partitions <- calculatePartitions(selfTdi, tdi)
+      _ <- trustDataInternalCache.modify(_ => (tdi, ()))
+      _ <- partitionCache.modify(_ => (partitions, ()))
+    } yield ()
+
+  private def cachedPartitionsCoverAllPeers: F[Boolean] =
+    for {
+      peerIds <- cluster.getPeerInfo.map(_.keySet) // TODO: Should we filter out offline nodes?
```

Comment:
is it about filtering out after partitioning or before?

Reply:
Before. I am not sure if we should pass all the peers as input to Partitioner or only online ones

Reply:
In general we should ask that question to all the places where `cluster.getPeerInfo.map(_.keySet)` is used